auctions/dutch-auction/src/main.sw

```diff
@@ -0,0 +1,280 @@
+
+        log(CancelledAuctionEvent {
+            id: auction_id, 
+        });
+    }
+
+    /// Returns the auction for any given auction_id
+    /// Panics when -
+    ///     1. auction_id is 0 or higher than storage.auction_count
+    fn auction(auction_id: u64) -> Auction {
+        validate_id(auction_id);
+        storage.auctions.get(auction_id)
+    }
+
+    /// Changes the bidding asset in the given auction
+    /// Panics when -
+    ///     1. auction_id is 0 or higher than storage.auction_count
+    ///     2. msg_sender is not the beneficiary of the auction
+    ///     3. auction has already ended
+    fn change_asset(new_asset: ContractId, auction_id: u64) {
+        validate_id(auction_id);
+        let mut auction = storage.auctions.get(auction_id);
+
+        // Only the beneficiary can change the bidding asset
+        require(eq_identity(sender_indentity(), auction.beneficiary), UserError::SenderNotBeneficiary);
+        // Cant edit an ended auction
+        require(!auction.ended, TimeError::AuctionAlreadyEnded);
+
+        auction.asset_id = new_asset;
+
+        storage.auctions.insert(auction_id, auction);
+    }
+
+    /// Changes the beneficiary of the given auction
+    /// Panics when -
+    ///     1. auction_id is 0 or higher than storage.auction_count
+    ///     2. msg_sender is not the beneficiary of the auction
+    ///     3. auction has already ended
+    fn change_beneficiary(new_beneficiary: Identity, auction_id: u64) {
+        validate_id(auction_id);
+        let mut auction = storage.auctions.get(auction_id);
+
+        // Only the beneficiary can change the beneficiary
+        require(eq_identity(sender_indentity(), auction.beneficiary), UserError::SenderNotBeneficiary);
+        // Cant edit an ended auction
+        require(!auction.ended, TimeError::AuctionAlreadyEnded);
+
+        auction.beneficiary = new_beneficiary;
+        /// WARNING: This needs to be changed to a pop to a vec instead of just replacing the contents of the array
+        storage.active_auctions_by_identity.insert(sender_indentity(), [0]);
+        /// WARNING: This needs to be changed to a push to a vec instead of just replacing the contents of the array
+        storage.active_auctions_by_identity.insert(new_beneficiary, [auction_id]);
+
+        storage.auctions.insert(auction_id, auction);
+    }
+
+    /// Returns the active auctions of which the identity is the beneficiary of
+    fn active_auctions_by_identity(identity_to_check: Identity) -> [u64;
+    1] {
+        storage.active_auctions_by_identity.get(identity_to_check)
+    }
+}
+
+/// This function is called whenever a winning bid is recieved.
+fn on_win(auction: Auction, winning_amount: u64) {
+    // Add custom logic for winning the auction here
+    transfer_to_identity(winning_amount, auction.asset_id, auction.beneficiary);
+}
+
+/// Calculates the current price of a given auction
+fn calculate_price(auction_id: u64) -> u64 {
```

Comment:
Can you move these private functions into `utils.sw` and import them?
Note that you cannot pass storage in directly into the function nor can you use it in a library so you must pass in the auction itself as an argument, likewise the auction count